File: java/src/test/java/money_problem/domain/PortfolioTest.java
package money_problem.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static money_problem.domain.Currency.*;
import static money_problem.domain.DomainUtility.*;
import static org.assertj.vavr.api.VavrAssertions.assertThat;

class PortfolioTest {
    private Bank bank;

    @BeforeEach
    void setup() {
        bank = Bank.withExchangeRate(EUR, USD, 1.2)
                .addExchangeRate(USD, KRW, 1100);
    }

    @Test
    @DisplayName("5 USD + 10 USD = 15 USD")
    void shouldAddMoneyInTheSameCurrency() {
        var portfolio = portfolioWith(
                dollars(5),
                dollars(10)
        );

        assertThat(portfolio.evaluate(bank, USD))
                .containsOnRight(dollars(15));
    }

    @Test
    @DisplayName("5 USD + 10 EUR = 17 USD")
    void shouldAddMoneyInDollarsAndEuros() {
        var portfolio = portfolioWith(
                dollars(5),
                euros(10)
        );

        assertThat(portfolio.evaluate(bank, USD))
                .containsOnRight(dollars(17));
    }

    @Test
    @DisplayName("1 USD + 1100 KRW = 2200 KRW")
    void shouldAddMoneyInDollarsAndKoreanWons() {
        var portfolio = portfolioWith(
                dollars(1),
                koreanWons(1100)
        );

        assertThat(portfolio.evaluate(bank, KRW))
                .containsOnRight(koreanWons(2200));
    }

    @Test
    @DisplayName("5 USD + 10 EUR + 4 EUR = 21.8 USD")
    void shouldAddMoneyInDollarsAndMultipleAmountInEuros() {
        var portfolio = portfolioWith(
                dollars(5),
                euros(10),
                euros(4)
        );

        assertThat(portfolio.evaluate(bank, USD))
                .containsOnRight(dollars(21.8));
    }

    @Test
    @DisplayName("Return a failure result in case of missing exchange rates")
    void shouldReturnAFailingResultInCaseOfMissingExchangeRates() {
        var portfolio = portfolioWith(
                euros(1),
                dollars(1),
                koreanWons(1)
        );

        assertThat(portfolio.evaluate(bank, EUR))
                .containsOnLeft("Missing exchange rate(s): [USD->EUR],[KRW->EUR]");
    }
}

File: c#/money-problem.Tests/PortfolioTest.cs
using System.Linq;
using FluentAssertions.LanguageExt;
using money_problem.Domain;
using Xunit;

namespace money_problem.Tests;

public class PortfolioTest
{
    private readonly Bank bank;

    public PortfolioTest() =>
        this.bank = DomainUtility.WithExchangeRates(
            Bank.WithPivotCurrency(Currency.EUR),
            DomainUtility.CreateExchangeRate(Currency.USD, 1.2),
            DomainUtility.CreateExchangeRate(Currency.KRW, 1344));

    [Fact(DisplayName = "5 USD + 10 EUR = 17 USD")]
    public void Add_ShouldAddMoneyInDollarAndEuro() =>
        PortfolioWith(5d.Dollars(), 10d.Euros())
            .Evaluate(this.bank, Currency.USD)
            .Should()
            .Be(17d.Dollars());

    [Fact(DisplayName = "1 USD + 1100 KRW = 2220 KRW")]
    public void Add_ShouldAddMoneyInDollarAndKoreanWons() =>
        PortfolioWith(1d.Dollars(), 1100d.KoreanWons())
            .Evaluate(this.bank, Currency.KRW)
            .Should()
            .Be(2220d.KoreanWons());

    [Fact(DisplayName = "5 USD + 10 EUR + 4 EUR = 21.8 USD")]
    public void Add_ShouldAddMoneyInDollarsAndMultipleAmountInEuros() =>
        PortfolioWith(5d.Dollars(), 10d.Euros(), 4d.Euros())
            .Evaluate(this.bank, Currency.USD)
            .Should()
            .Be(21.8.Dollars());

    [Fact(DisplayName = "Throws a MissingExchangeRatesException in case of missing exchange rates")]
    public void Add_ShouldThrowAMissingExchangeRatesException()
    {
        PortfolioWith(1d.Euros(), 1d.Dollars(), 1d.KoreanWons())
            .Evaluate(Bank.WithPivotCurrency(Currency.EUR), Currency.EUR)
            .Should()
            .Be("Missing exchange rate(s): [USD->EUR],[KRW->EUR]");
    }

    [Fact(DisplayName = "5 USD + 10 USD = 15 USD")]
    public void Add_ShouldAddMoneyInTheSameCurrency() =>
        PortfolioWith(5d.Dollars(), 10d.Dollars())
            .Evaluate(this.bank, Currency.USD)
            .Should()
            .Be(15d.Dollars());

    private static Portfolio PortfolioWith(params Money[] moneys) =>
        moneys.Aggregate(new Portfolio(), (portfolio, money) => portfolio.Add(money));
}